File: InterviewPrep/Leetcode/JS/arrays/53MaximumSubarray.js
// Kadane's Algorithm, maximum subarray sum




/////////sol1/////////


var maxSubArray = function (nums) {
    let max_sum = -Infinity;
    let curr_sum = 0;
    for (let i = 0; i < nums.length; i++) {
        curr_sum = curr_sum + nums[i];
        if (curr_sum > max_sum) {
            max_sum = curr_sum;
        }
        if (curr_sum < 0) {
            curr_sum = 0
        }
    }
    return max_sum

};



////////////sol2//////////

var maxSubArray = function (nums) {
    let max_sum = -Infinity;
    let curr_sum = 0;
    for (let i = 0; i < nums.length; i++) {
        // updating the curr_sum till ow.. it can be current element or curr_sum+current element
        curr_sum = Math.max(curr_sum + nums[i], nums[i]);
        if (curr_sum > max_sum) {
            max_sum = curr_sum;
        }

    }
    return max_sum

};


TC:O(n)
SC:O(1)

File: InterviewPrep/Leetcode/Kadane'sAlgorithm/53MaximumSubarray.cpp
Given an integer array nums, find the 
subarray
 with the largest sum, and return its sum.

 

Example 1:

Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
Output: 6
Explanation: The subarray [4,-1,2,1] has the largest sum 6.
Example 2:

Input: nums = [1]
Output: 1
Explanation: The subarray [1] has the largest sum 1.
Example 3:

Input: nums = [5,4,-1,7,8]
Output: 23
Explanation: The subarray [5,4,-1,7,8] has the largest sum 23.
 

Constraints:

1 <= nums.length <= 105
-104 <= nums[i] <= 104





////////////////////sol////////////////////


class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int max_end_here=0,max_so_far=INT_MIN;
        for(int i=0;i<nums.size();i++){
            max_end_here+=nums[i];
            if(max_so_far<max_end_here)
            max_so_far=max_end_here;
            if(max_end_here<0)
            max_end_here=0;
        }
        return max_so_far;
        
    }
};





//Note:
// The idea of Kadane’s algorithm is to maintain a variable max_ending_here that stores the maximum sum contiguous 
// subarray ending at current index and a variable max_so_far stores the maximum sum of contiguous subarray found so far,
// Everytime there is a positive-sum value in max_ending_here compare it with max_so_far and update max_so_far 
// if it is greater than max_so_far.